src/bosh_azure_cpi/lib/cloud/azure/vm_manager.rb

```diff
@@ -474,62 +481,41 @@ def get_availability_set_name(resource_pool, env)
       availability_set_name
     end
 
-    def create_availability_set(resource_group_name, availability_set_params)
+    def get_or_create_availability_set(resource_group_name, availability_set_params)
       availability_set_name = availability_set_params[:name]
-      availability_set = @azure_client2.get_availability_set_by_name(resource_group_name, availability_set_name)
-
-      if availability_set.nil?
-        @logger.info("create_availability_set - the availability set `#{availability_set_name}' doesn't exist. Will create a new one.")
-      else
+      availability_set = nil
+      flock("#{CPI_LOCK_PREFIX_AVAILABILITY_SET}-#{availability_set_name}", File::LOCK_EX) do
+        availability_set = @azure_client2.get_availability_set_by_name(resource_group_name, availability_set_name)
+        if availability_set.nil?
+          @logger.info("The availability set `#{availability_set_name}' doesn't exist. Will create a new one.")
+          @azure_client2.create_availability_set(resource_group_name, availability_set_params)
+          availability_set = @azure_client2.get_availability_set_by_name(resource_group_name, availability_set_name)
         # In some regions, the location of availability set is case-sensitive, e.g. CanadaCentral instead of canadacentral.
-        if availability_set[:location].downcase != availability_set_params[:location].downcase
-          cloud_error("create_availability_set - the availability set `#{availability_set_name}' already exists, but in a different location `#{availability_set[:location].downcase}' instead of `#{availability_set_params[:location].downcase}'. Please delete the availability set or choose another location.")
-        else
-          if !@use_managed_disks && availability_set[:managed]
-            cloud_error("create_availability_set - the availability set `#{availability_set_name}' already exists. It's not allowed to update it from managed to unmanaged.")
-          elsif @use_managed_disks && !availability_set[:managed]
-            @logger.info("create_availability_set - the availability set `#{availability_set_name}' exists, but it needs to be updated from unmanaged to managed.")
-            availability_set_params.merge!({
-              :platform_update_domain_count => availability_set[:platform_update_domain_count],
-              :platform_fault_domain_count  => availability_set[:platform_fault_domain_count],
-              :managed                      => true
-            })
-          else
-            @logger.info("create_availability_set - the availability set `#{availability_set_name}' exists. No need to update.")
-            return availability_set
-          end
-        end
-      end
-
-      mutex = FileMutex.new("#{CPI_LOCK_PREFIX_AVAILABILITY_SET}-create-#{availability_set_name}", @logger)
-      begin
-        if mutex.lock
+        elsif !availability_set[:location].casecmp(availability_set_params[:location]).zero?
```

Comment:
is this change related to the issue customer reported?

Reply:
yes, there can be the same error in create_vm.

Reply:
I mean the changes like:
1. @logger.info("The availability set `#{availability_set_name}' doesn't exist. Will create a new one.")
2. casecmp
3. elsif  adjustment.

Reply:
I ask this is because basicially the hotfix should contains as minor code change as possible.

Reply:
Got your point, this function is part of #414 (except the logger info) which was already tested, and using a tested function is safer than hijack it.